components/layout_2020/query.rs

```diff
@@ -446,45 +452,70 @@ fn process_offset_parent_query_inner(
-                                // Again, take the *first* associated CSS layout box.
-                                let padding_box_corner = fragment
-                                    .padding_rect()
-                                    .to_physical(fragment.style.writing_mode, containing_block)
-                                    .origin
-                                    .to_vector() +
-                                    containing_block.origin.to_vector();
-                                let padding_box_corner = padding_box_corner.to_untyped();
-                                Some(padding_box_corner)
-                            } else {
-                                None
-                            }
-                        },
-                        Fragment::AbsoluteOrFixedPositioned(_) |
-                        Fragment::Text(_) |
-                        Fragment::Image(_) |
-                        Fragment::IFrame(_) |
-                        Fragment::Positioning(_) => None,
-                    }
-                })
-                .unwrap()
-        })
-        // "If the offsetParent of the element is null," subtract zero in the
-        // following step.
-        .unwrap_or(Vector2D::zero());
+    let offset_parent_padding_box_corner =
+        if let Some(offset_parent_node_address) = node_offset_box.offset_parent_node_address {
+            // Although the spec (https://www.w3.org/TR/cssom-view-1/#extensions-to-the-htmlelement-interface)
+            // says that offsetTop/offsetLeft are always relative to the padding box of the offsetParent.
+            // In practice this is not true in major browsers in the case that the offsetParent is the body
+            // element and the body element is position:static. In that case offsetLeft/offsetTop are computed
+            // relative to the root nodes border box.
+            if node_offset_box.is_static_body_element {
+                let containing_block = &fragment_tree.initial_containing_block;
+                match &(*fragment_tree.root_fragments[0].borrow()) {
+                    Fragment::Box(fragment) | Fragment::Float(fragment) => {
+                        // Again, take the *first* associated CSS layout box.
+                        let border_box_corner = fragment
+                            .border_rect()
+                            .to_physical(fragment.style.writing_mode, containing_block)
+                            .origin
+                            .to_vector() +
+                            containing_block.origin.to_vector();
+                        border_box_corner.to_untyped()
+                    },
+                    Fragment::AbsoluteOrFixedPositioned(_) |
+                    Fragment::Text(_) |
+                    Fragment::Image(_) |
+                    Fragment::IFrame(_) |
+                    Fragment::Positioning(_) => unreachable!(),
```

Comment:
This is not unreachable, e.g. this crashes

```html
<!DOCTYPE html>
<html style="position: absolute">
<body id="body" onload="recompute()" style="border: solid">
  <div id="target" style="height: 40px; width: 40px; background: red;"></div>
  <div id="output" style="text-wrap: nowrap"></div>
  <button onclick="recompute()">recompute</button>
  <script type="text/javascript">
    function recompute() {
      let targetNode = document.getElementById("target");
      let outputNode = document.getElementById("output");
      outputNode.innerHTML = `
        offsetParent: ${targetNode.offsetParent.id}<br />
        offsetTop: ${targetNode.offsetTop}<br />
        offsetLeft: ${targetNode.offsetLeft}<br />
      `;
    }
  </script>
</body>
```

Reply:
I have fixed this particular crash. Not sure if the root element can be any of the other variants.